lib/load-config.js

```diff
@@ -22,30 +22,131 @@ const CONFIG_FILE_NAMES = Object.freeze([
   'nyc.config.cjs'
 ])
 
-function loadConfigFile (path) {
-  let config = {}
-  const jsExts = ['.js', '.cjs']
-  const ymlExts = ['.yml', '.yaml']
-  const fileName = basename(path)
+const EXTS_JS_VALIDATOR = /\.c?js$/i
+const EXTS_YAML_VALIDATOR = /\.ya?ml$/i
 
-  const ext = extname(path).toLowerCase()
-  if (jsExts.includes(ext)) {
-    config = require(path)
-  } else if (ymlExts.includes(ext)) {
-    config = require('js-yaml').load(readFileSync(path, 'utf8'))
-  } else if (ext === '.json' || fileName.slice(-2) === 'rc') {
-    config = JSON.parse(readFileSync(path, 'utf8'))
+class UnsupportedFileTypeError extends Error {
+  constructor (path) {
+    super(
+      `Unsupported file type ${extname(
+        path
+      ).toLowerCase()} while reading file ${path}`
+    )
+    this.name = UnsupportedFileTypeError.name
+  }
+}
+
+class ConfigParsingError extends Error {
+  constructor (path, errorDetails, originalError) {
+    super(
+      `Error loading configuration from file ${path}: ${errorDetails}` +
+        (typeof originalError === 'object' && originalError instanceof Error
+          ? ` Original error: ${originalError.message}`
+          : '')
+    )
+
+    this.name = UnsupportedFileTypeError.name
+
+    if (originalError instanceof Error) {
+      this.stack = originalError.stack
+    }
   }
+}
+
+const NO_EXPORTS = Symbol('no exports')
+
+function loadConfigFile (path, _di) {
```

Comment:
Question: What does the second parameter di?  I assume an array or an object. Does the di stand for dependency injection?  I see it called a great deal in the test cases.  I am concerned this might get flagged in another peer review.

I did not address in the review branch.

Reply:
Correct, It's what is known to me as "poor man's Dependency Injection" - its purpose would be clearer in TypeScript, where the signature would look like `function loadConfigFile (path: string, _di?: {readFile: (path: string) => string, readJs: (path: string) => string}) {` - it's ONLY to be used by the test system to allow stubbing the side-loaded inputs to the function.

I note that while I call it "poor man's DI", it's the ONLY option I've found when it comes to DI for functions in JavaScript. All DI libraries I've found only work with classes.

Reply:
Since this is JavaScript and not TypeScript we could get away with altering the code to use the following pattern:

```javascript
function loadConfigFile (path) {
  const di = {
    readFile: (path) => readFileSync(path, 'utf8'),
    readJs: (path) => require(path),
    ...arguments[1] // Second argument is only ever provided as dependency injection from the test suites.
  }
```

Reply:
Issues with this implementation of Dependency injection aka di.

In most implementations of di, implementation files don't call the constructor.  It's written in a general way and implemented in the API.  It's decoupled.  Overall the di pattern does extend the constructor to add behavior.  Almost like writing a class function and requiring an interface.  This extension can cause additional errors and it can be troublesome to debug.  It's a free-for-all all on the definition of a class, instead of extending a class definition in another file.  

When I was first introduced to the di, my first thought was, "This is an alternative to the factory pattern".  Some accuse di of being a golden hammer, which is one of the warnings of the book titled ["Design Patterns"](https://www.amazon.com/gp/product/0201633612/ref=as_li_tl?ie=UTF8&camp=1789&creative=390957&creativeASIN=0201633612&linkCode=as2&tag=triatcraft-20&linkId=XRGUDJCGWC6AJNZM)

Could it be possible to rework this class and the test cases not to use this pattern?

Reply:
Class? There's no class here, only a function. Unless you are conceptualizing JavaScript functions as classes with implicit constructors, which they CAN be, but that concept leads to difficult to maintain code - and often requires a complete rewrite to an actual `class` when the project matures further.

In this case I simply wanted to be able to inject alternative responses to the side-loaded inputs to this file, the calls to reading from disk. There are two approaches I can see: the first is to do as I've already demonstrated. The second is to add optional parameters, a la

```javascript
function loadConfigFile (path, readFile = (path) => readFileSync(path, 'utf8'), readJs = require(path)) {
```

But this means that if I want to write an integration test that needs to override `readJs` but leave `readFile` defined with the default I'm stuck as I cannot do that. There's complex ways around that, but the simplest approach is to pass an object, e.g.

```javascript
function loadConfigFile (path, di = { readFile: (path) => readFileSync(path, 'utf8'), readJs: require(path) }) {
```

But that has the same problem: if I want to write an integration test that stubs one but not the other I'm stuck. So instead I allow for either to be overridden by using the `{ defaultValuesHere, ..._di}` syntax as I showed in this PR.

As to off-the-shelf DI, let's go down the rabbit hole: let's say we redid this as a class just so we could use off-the-shelf DI. We'd have to create classes that wrap the functionality we are wanting to inject. We'd then have to register them to the DI library, often using JS decorators, and then register the class that wraps `loadConfigFile` ,again most likely using JS decorators. Oops, JS support for decorators doesn't exist in Node14 IIRC, plus a lot of work to solve a problem when a simpler pattern exists.

Reply:
To respond directly to your question: Yes, this could be reworked to not pass in alternative implementations of those two functions. There are two approaches I'm familiar with:

1. Use Sinon's ability to mutate the global namespace the functions come from and thus stub any call to `require` or `readFileSync`. However mutating global anything in a test suite is an anti-pattern, plus it is explicitly disallowed in ESM. So I no longer use this technique. Plus I'm not certain that stubbing `require` is even possible or wanted - which would lead to creating another file which implements a function that calls require - a pointless waste of time IMO.
2. Do as you've done elsewhere and create files on disk for it to read. This results in an _integration_ test not a _unit_ test and as a consequence makes a test that should have taken a couple of milliseconds at best take an order or two of magnitude longer due to the IO delay - plus it potentially results in random test failures if the disk has issues. Since other test suites are indeed exercising that path of reading files, I figured I didn't need to write another integration test, I just needed to write a unit test - and to me that means that I have direct programmatic control of every input tot he function under test, not indirectly by having it also hit the filesystem.

Reply:
Thank you for all your notes about this concern today.  I am looking more into the topic, and I might have a new understanding of pure di or poor man's di.  Give me a bit more time to review this topic you presented.  I will get back to this next week.

Reply:
When I started learning about DI myself I [read an article](https://dev.to/vovaspace/dependency-injection-in-typescript-4mbf) that began explaining about the need of DI by using a simple parameter-based injection of a dependency into a function. But the author then immediately jumped to classes and then went to his favorite DI Container tool without giving any other solutions for DI on functions. After a year or so of using various hacks, I found out at least one reason why: all these DI Container systems use decorators, and decorators cannot be used on functions. As I continued to dig I found that a LOT of people were using one-off techniques much like I've done here, though often not as advanced as they've not had to support the use-cases I've met while using this DI pattern daily at work.